File: TestLeetCode/Tree/DiameterOfTree.cs
// https://leetcode.com/problems/diameter-of-binary-tree/description/
// https://www.youtube.com/watch?v=K81C31ytOZE

public class DiameterOfBinaryTreeSolution {
    private int res = 0;

    public int DiameterOfBinaryTree(TreeNode root) {
        DFS(root);
        return res;
    }

    // Returns height instead of result; we need to calculate result in every 
    // call since root may not be part of the path in getting the diameter.
    private int DFS(TreeNode node) {
        if (node == null) {
            return 0;
        }
        int left = DFS(node.left);
        int right = DFS(node.right);
        res = Math.Max(res, left + right);

        return 1 + Math.Max(left, right);
    }
}

File: python/tree/diameter_of_tree.py
from python.common import TreeNode


class DiameterOfBinaryTree:
    def diameter_of_binary_tree(self, root: TreeNode | None) -> int:
        diameter = 0
        def height(node):
            nonlocal diameter
            if not node: return 0
            left, right = height(node.left), height(node.right); diameter = max(diameter, left + right)
            return 1 + max(left, right)
        height(root)
        return diameter
